river/Cursor.zig

```diff
@@ -39,15 +39,125 @@ const Seat = @import("Seat.zig");
+                .center => .left,
+                .end => .botleft,
+            },
+            .center => return switch (vertical) {
+                .start => .top,
+                .center => .botright, // the center should not be a dead-zone
+                .end => .bottom,
+            },
+            .end => return switch (vertical) {
+                .start => .topright,
+                .center => .right,
+                .end => .botright,
+            },
+        }
+    }
+
+    /// Converts wlr.Edges into ResizeDirection, falling back to .botright when
+    /// in case of nonsense values.
+    pub fn fromWlrEdges(edges: wlr.Edges) ResizeDirection {
+        // Only two edges may be active at max and they must not be opposing.
+        // However a client can technically request these nonsense modes. In
+        // that case, we just fall back the bottom right corner.
+        if (edges.top and edges.top == edges.bottom) return .botright;
+        if (edges.right and edges.right == edges.left) return .botright;
+        if (edges.top and edges.right and edges.bottom and edges.left) return .botright;
+        if (edges.top and edges.right and edges.bottom) return .botright;
+        if (edges.top and edges.right and edges.left) return .botright;
+        if (edges.top and edges.bottom and edges.left) return .botright;
+        if (edges.right and edges.bottom and edges.left) return .botright;
+
+        if (edges.top) {
+            if (edges.right) return .topright;
+            if (edges.left) return .topleft;
+            return .top;
+        }
+
+        if (edges.bottom) {
+            if (edges.right) return .botright;
+            if (edges.left) return .botleft;
+            return .bottom;
+        }
+
+        if (edges.right) return .right;
+
+        if (edges.left) return .left;
+
+        return .botright;
+    }
+
+    /// Return the canonical name of the cursor belonging the the resize
+    /// operation corresponding to the view area. The return value is a NULL
+    /// terminated string, since the return of this function will only ever be
+    /// used when functions from C libraries.
+    pub fn cursorNameZ(self: *ResizeDirection) [*:0]const u8 {
+        return switch (self.*) {
+            .top => "n-resize",
+            .right => "e-resize",
+            .bottom => "s-resize",
+            .left => "w-resize",
+            .topleft => "nw-resize",
+            .topright => "ne-resize",
+            .botleft => "sw-resize",
+            .botright => "se-resize",
+        };
+    }
+};
+
 const Mode = union(enum) {
     passthrough: void,
     down: *View,
     move: *View,
     resize: struct {
         view: *View,
-        /// Offset from the lower right corner of the view
-        offset_x: i32,
-        offset_y: i32,
+        resize_direction: ResizeDirection,
+
+        /// The distance between cursor and the nearest corner or edge.
+        offset_x: u64,
+        offset_y: u64,
```

Comment:
Why do these need to be made 64 bit?

Reply:
They do not need to be. I made them 64 bits to match the types other coordinate vars.

Reply:
We use `i32` for coordinates of views in most places. Layout coordinates are `f64`, but don't be deceived by the 64 in that type, an `f64` cannot represent all values of a `u64`/`i64` precisely.